File: java/src/main/java/tech/qmates/openchat/domain/usecase/RegisterUserUsecase.java
package tech.qmates.openchat.domain.usecase;

import tech.qmates.openchat.domain.entity.User;
import tech.qmates.openchat.domain.repository.UserRepository;

public class RegisterUserUsecase {

    private final UserRepository userRepository;

    public RegisterUserUsecase(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public void run(String username) throws UsernameAlreadyInUseException {
        if(userRepository.isUsernameAlreadyUsed(username))
            throw new UsernameAlreadyInUseException(username);

        this.userRepository.store(new User(username));
    }

    public static class UsernameAlreadyInUseException extends Exception {
        public UsernameAlreadyInUseException(String username) {
            super("Username [" + username + "] already in use!");
        }
    }
}

File: typescript/src/domain/usecases/RegisterUserUseCase.ts
import { newUserToRegister } from '../entities/User';
import UserRepository from "../repositories/UserRepository";

export default class RegisterUserUseCase {
  private userRepository: UserRepository;

  constructor(userRepository: UserRepository) {
    this.userRepository = userRepository
  }

  run(username: string, password: string, userAbout: string): string {
    if (this.userRepository.isUsernameAlreadyUsed(username))
      throw new UsernameAlreadyInUseError(username)

    const userToStore = newUserToRegister(username, password, userAbout)
    this.userRepository.store(userToStore)

    return userToStore.id
  }
}

export class UsernameAlreadyInUseError extends Error {
  constructor(username: string) {
    super(`Username ${username} already in use!`)
    Object.setPrototypeOf(this, UsernameAlreadyInUseError.prototype);
  }
}